builds.sh

```diff
@@ -1,228 +1,19 @@
-					.from
-				end
-			')"
-
-			# if "$lookup" is a valid/known sourceId, we should look up the (pre-resolved) imageId for it
-			if [ -n "$lookup" ] && [ -n "${sources["$lookup"]:+x}" ]; then
-				resolved="${sourceArchResolved["$lookup-$arch"]:-}"
-			elif [ -n "$lookup" ]; then
-				if [ -z "${imageArchResolved["$lookup"]:+x}" ]; then
-					remoteArches="$(_resolveRemoteArch "$lookup")"
-					imageArchResolved["$lookup"]="$remoteArches"
-				fi
-
-				resolved="$(_resolve "$lookup" "$arch" "${imageArchResolved["$lookup"]}")"
-			else
-				resolved=
-			fi
-			: "${resolved:=null}"
-
-			buildIdParts="$(
-				jq <<<"$buildIdParts" -c \
-					--argjson parent "$parent" \
-					--argjson resolved "$resolved" \
-					'
-						.resolvedParents[$parent.from] = $resolved
-						| .parents[$parent.from] = $resolved.manifest?.desc?.digest?
-					'
-			)"
-
-			if [ "$resolved" = 'null' ] || [ -z "$resolved" ]; then
-				(( missingParents++ )) || :
-			fi
-		done
-
-		# if we're missing *any* parents, we cannot have a buildId
-		if [ "$missingParents" = 0 ]; then
-			buildIdJson="$(jq <<<"$buildIdParts" -c 'del(.resolvedParents)')" # see notes above (where buildIdParts is first defined)
-			buildId="$(sha256sum <<<"$buildIdJson" | cut -d' ' -f1)" # see notes above (where buildIdParts is first defined)
-			printf >&2 '%s\n' "$buildId"
-
-			img="$BASHBREW_STAGING_TEMPLATE"
-			img="${img//BUILD/$buildId}"
-			[ "$img" != "$BASHBREW_STAGING_TEMPLATE" ] # BUILD is required, for proper uniqueness (ARCH is optional)
-			img="${img//ARCH/$arch}"
-
-			if [ -z "${imageArchResolved["$img"]:+x}" ]; then
-				remoteArches="$(_resolveRemoteArch "$img")"
-				imageArchResolved["$img"]="$remoteArches"
-			fi
-
-			resolved="$(_resolve "$img" "$arch" "${imageArchResolved["$img"]}")"
-			: "${resolved:=null}"
-			sourceArchResolved["$sourceId-$arch"]="$resolved"
-			builds="$(
-				jq <<<"$builds" \
-					--argjson source "$obj" \
-					--argjson buildIdParts "$buildIdParts" \
-					--arg buildId "$buildId" \
-					--arg img "$img" \
-					--argjson resolved "$resolved" \
-					'
-						# TODO error out if $buildId already exists somehow (should be impossible)
-						.[$buildId] = {
-							buildId: $buildId,
-							build: ({
-								img: $img,
-								resolved: $resolved,
-							} + $buildIdParts),
-							source: (
-								$source
-								| .arches |= (
-									(keys_unsorted - [ $buildIdParts.arch ]) as $otherArches
-									| del(.[$otherArches[]])
-								)
-							),
-						}
-					'
-			)"
-		else
-			printf >&2 'not yet!\n'
-		fi
-	done
-done
-jq <<<"$builds" .
+dir="$(dirname "$BASH_SOURCE")"
+dir="$(readlink -ve "$dir")"
+if [ "$dir/builds.go" -nt "$dir/builds" ] || [ "$dir/om/om.go" -nt "$dir/builds" ] || [ "$dir/.go-env.sh" -nt "$dir/builds" ]; then
+	{
+		echo "building '$dir/builds' from 'builds.go'"
+		"$dir/.go-env.sh" go build -v -o builds builds.go
+		ls -l "$dir/builds"
+	} >&2
+fi
```

Comment:
`Makefile`?

Reply:
I thought about it, but I'm not sure it actually makes this much simpler since it still has to list all the files that should cause the binary to rebuild, and it's really creating the container (via `.go-env.sh`) that I was trying to optimize not doing unless necessary, not the `go build` (which is otherwise _really_ fast and I've have no problem just running unconditionally every time).  I also don't want to encode an assumption of `make` being available on the host in scripts like this (as a general rule), so even with a `Makefile` I'd still be invoking it via `.go-env.sh` and be back to the same problem. :see_no_evil:

Instead of maintaining another file that then splits the logic for these things, I'd probably instead go the other direction and just pay the cost of the container overhead on every startup (since it should typically only be _real_ overhead the first time, when it has to pull the image).